mmv1/third_party/terraform/fwvalidators/framework_validators.go

```diff
@@ -115,3 +117,109 @@ func (v nonEmptyStringValidator) ValidateString(ctx context.Context, request val
 func NonEmptyStringValidator() validator.String {
 	return nonEmptyStringValidator{}
 }
+
+func StringSet(d basetypes.SetValue) []string {
+
+	StringSlice := make([]string, 0)
+	for _, v := range d.Elements() {
+		StringSlice = append(StringSlice, v.(basetypes.StringValue).ValueString())
+	}
+	return StringSlice
+}
+
+// Define the possible service account name patterns
+var serviceAccountNamePatterns = []string{
+	`^.+@.+\.iam\.gserviceaccount\.com$`,                     // Standard IAM service account
+	`^.+@developer\.gserviceaccount\.com$`,                   // Legacy developer service account
+	`^.+@appspot\.gserviceaccount\.com$`,                     // App Engine service account
+	`^.+@cloudservices\.gserviceaccount\.com$`,               // Google Cloud services service account
+	`^.+@cloudbuild\.gserviceaccount\.com$`,                  // Cloud Build service account
+	`^service-[0-9]+@.+-compute\.iam\.gserviceaccount\.com$`, // Compute Engine service account
+}
+
+// Create a custom validator for service account names
+type ServiceAccountNameValidator struct{}
+
+func (v ServiceAccountNameValidator) Description(ctx context.Context) string {
+	return "value must be a valid service account email address"
+}
+
+func (v ServiceAccountNameValidator) MarkdownDescription(ctx context.Context) string {
+	return v.Description(ctx)
+}
+
+func (v ServiceAccountNameValidator) ValidateString(ctx context.Context, req validator.StringRequest, resp *validator.StringResponse) {
+	if req.ConfigValue.IsNull() || req.ConfigValue.IsUnknown() {
+		return
+	}
+
+	value := req.ConfigValue.ValueString()
+	valid := false
+	for _, pattern := range serviceAccountNamePatterns {
+		if matched, _ := regexp.MatchString(pattern, value); matched {
+			valid = true
+			break
+		}
+	}
+
+	// Check for empty string
+	if value == "" {
+		resp.Diagnostics.AddError("Invalid Service Account Name", "Service account name must not be empty")
+		return
+	}
+
+	if !valid {
+		resp.Diagnostics.AddAttributeError(
+			req.Path,
+			"Invalid Service Account Name",
+			"Service account name must match one of the expected patterns for Google service accounts",
+		)
+	}
+}
+
+// Create a custom validator for duration
+type DurationValidator struct {
```

Comment:
I'm wondering if this validator can be named more specifically. For example there's a `NonNegativeDurationValidator` in the code base and it's more clear what it does from the name versus `DurationValidator`. Can you think of a more descriptive name that matches what this new validator does?

Reply:
It make sense to include bound in the name something like `BoundedDuration`.

I don't think we need to include `NonNegative` if we're specifying that this Duration is Bounded based on what MinDuration and MaxDuration are set to.

Reply:
Thanks for updating from `DurationValidator` to `BoundedDuration` - that gives more context about what the validator does.

>I don't think we need to include NonNegative if we're specifying that this Duration is Bounded based on what MinDuration and MaxDuration are set to.

Agreed  - `NonNegativeDurationValidator` is validator that's already in the code base that's an example of a more descriptive name. Sorry I wasn't clear